Firestore/core/src/firebase/firestore/core/query.cc

```diff
@@ -45,18 +62,46 @@ bool Query::IsDocumentQuery() const {
          filters_.empty();
 }
 
+const FieldPath* Query::InequalityFilterField() const {
+  for (const auto& filter : filters_) {
+    if (filter->IsInequality()) {
+      return &filter->field();
+    }
+  }
+  return nullptr;
+}
+
+bool Query::HasArrayContainsFilter() const {
+  for (const auto& filter : filters_) {
+    if (filter->type() == Type::kRelationFilter) {
+      const auto& relationFilter =
+          std::static_pointer_cast<RelationFilter>(filter);
```

Comment:
Nit: `static_pointer_cast` will create a copy of the `shared_ptr`, perhaps just call `get` and then cast the raw pointer?